File: tests/bench/crosslang/c/branch.c
/* Порт tests/bench/programs/BranchBench.eat 1:1 — непредсказуемые
 * ветвления, if/elif-диспетчер по битам loop-carried acc. Вывод обязан
 * совпасть байт-в-байт с EATLang-оригиналом при том же REPEAT. */
#include <stdint.h>
#include <stdio.h>

static const uint32_t REPEAT = 1;

static uint32_t work(uint32_t seed) {
    uint32_t acc = seed + 1u;
    for (uint32_t k = 0; k < 500u; k++) {
        for (uint32_t i = 0; i < 1000u; i++) {
            uint32_t t = (acc >> (i & 7u)) & 3u;
            if (t == 0u) {
                acc = (acc + i + 1u) % 65536u;
            } else if (t == 1u) {
                acc = (acc * 5u + 3u) % 65536u;
            } else if (t == 2u) {
                acc = (acc ^ (i & 4095u)) % 65536u;
            } else {
                acc = ((acc >> 1) | 1u) % 65536u;
            }
        }
    }
    return acc;
}

int main(void) {
    uint32_t acc = 7u;
    for (uint32_t r = 0; r < REPEAT; r++) {
        acc = work(acc);
    }
    printf("checksum %u\n", acc);
    return 0;
}

File: tests/bench/crosslang/rust/branch.rs
// Порт tests/bench/programs/BranchBench.eat 1:1 — непредсказуемые
// ветвления, if/elif-диспетчер по битам loop-carried acc. Вывод обязан
// совпасть байт-в-байт с EATLang-оригиналом при том же REPEAT.

const REPEAT: u32 = 1;

fn work(seed: u32) -> u32 {
    let mut acc: u32 = seed + 1;
    for _ in 0..500u32 {
        for i in 0..1000u32 {
            let t: u32 = (acc >> (i & 7)) & 3;
            if t == 0 {
                acc = (acc + i + 1) % 65536;
            } else if t == 1 {
                acc = (acc * 5 + 3) % 65536;
            } else if t == 2 {
                acc = (acc ^ (i & 4095)) % 65536;
            } else {
                acc = ((acc >> 1) | 1) % 65536;
            }
        }
    }
    acc
}

fn main() {
    let mut acc: u32 = 7;
    for _ in 0..REPEAT {
        acc = work(acc);
    }
    println!("checksum {}", acc);
}
